File: portlet/space/src/main/webapp/javascript/eXo/social/space/UISpaceSearch.js
/*
 * Copyright (C) 2003-2007 eXo Platform SAS.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Affero General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see<http://www.gnu.org/licenses/>.
 */

/**
 * Space search.
 * @class
 * @scope public
 */
function UISpaceSearch() {
	this.inputTextBoxObj = null;
	this.allSpaceName = null;
};

/**
 * Initialize ui controls when the form is loaded.
 * @ uicomponentId Id of current component
 * @ spaceNames All space name of current search.
 */
UISpaceSearch.prototype.onLoad = function(uicomponentId, spaceNames) {
	var DOMUtil = eXo.core.DOMUtil;
	var spaceSearch = document.getElementById(uicomponentId);
	var spaceSearchEl = DOMUtil.findDescendantById(spaceSearch, 'SpaceSearch');
	var defaultSpaceName = document.getElementById('defaultSpaceName').value;
	var defaultUIVal = "Space name";
	if (spaceSearchEl.value == defaultUIVal) spaceSearchEl.value = defaultSpaceName;
	(spaceSearchEl.value != defaultSpaceName) ? (spaceSearchEl.style.color = '#000000') : (spaceSearchEl.style.color = '#C7C7C7');
	this.inputTextBoxObj = spaceSearchEl;
	this.setAllSpaceName(spaceNames);
	// Initialize the input textbox
	this.initTextBox();
};

/**
 * Initialize the text-box control. 
 * @scope private.
 */
UISpaceSearch.prototype.initTextBox = function() {
	var searchEl = this.inputTextBoxObj;
	var defaultValue = document.getElementById('defaultSpaceName').value;
	var uiSpaceSearchObj = eXo.social.space.UISpaceSearch;
//	var suggestControlObj = eXo.social.space.AutoSuggestControl;
	var suggestControlObj = eXo.social.webui.UIAutoSuggestControl;
	// Turn off auto-complete attribute of text-box control
	searchEl.setAttribute('autocomplete','off');
	
	// Add focus event for control
	searchEl.onfocus = function() {
		searchEl.style.color="#000000";
		searchEl.focus();
		if (searchEl.value == defaultValue) {
			searchEl.value='';
		}
	}
	
	// Add blur event for control
	searchEl.onblur = function() {
		if ((searchEl.value.trim() == '') || (searchEl.value.trim() == defaultValue)) {
			searchEl.style.color="#C7C7C7";
			searchEl.value = defaultValue;
		} 
		suggestControlObj.hideSuggestions();
	}
	
	// Add keydown event for control
	searchEl.onkeydown = function(event) {
		var e = event || window.event;
		var keynum = e.keyCode || e.which;  
		  
		if(keynum == 13) { //Enter key
			suggestControlObj.hideSuggestions();
			uiSpaceSearchObj.submitSearchForm(searchEl);
		} else { // Other keys (up and down key)
			suggestControlObj.handleKeyDown(e);
		}
	}
	
	suggestControlObj.load(searchEl, uiSpaceSearchObj);
}

/**
 * Submit the search form.
 * @scope private.
 */
UISpaceSearch.prototype.submitSearchForm = function(searchEl) {
	var DOMUtil = eXo.core.DOMUtil;
	var UIForm = eXo.webui.UIForm;
	var searchForm = DOMUtil.findAncestorByClass(searchEl, 'UIForm');
	if (searchForm != null ) UIForm.submitForm(searchForm.id, 'Search', true);
}

/**
 * Set all space name to allSpaceNames variable.
 * @scope private.
 */
UISpaceSearch.prototype.setAllSpaceName = function(allName) {
	var allSpaceNames = allName.substring(1, allName.length-1);
	var allSN = allSpaceNames.split(',');
	var allNames = [];
	for (var i=0; i < allSN.length; i++) {
		(function(idx) {
			allNames.push(allSN[idx].trim());
		})(i);
	}
	this.allSpaceName = allNames;
};

/**
 * Request suggestions for the given autosuggest control. 
 * @scope protected
 * @param oAutoSuggestControl The autosuggest control to provide suggestions for.
 */
UISpaceSearch.prototype.requestSuggestions = function (oAutoSuggestControl /*:AutoSuggestControl*/) {
    var aSuggestions = [];
    var sTextboxValue = oAutoSuggestControl.textbox.value;
    
    if (sTextboxValue.length > 0){
    
        //convert value in textbox to lowercase
        var sTextboxValueLC = sTextboxValue.toLowerCase();
        
        //search for matching states
        for (var i=0; i < this.allSpaceName.length; i++) { 

            //convert state name to lowercase
            var sStateLC = this.allSpaceName[i].toLowerCase();
            
            //compare the lowercase versions for case-insensitive comparison
            if (sStateLC.indexOf(sTextboxValueLC) == 0) {
                //add a suggestion using what's already in the textbox to begin it                
            	aSuggestions.push(this.allSpaceName[i]);

                // Check if user type like the last suggestion in drop-down list.
                if ((this.allSpaceName[i].substring(sTextboxValue.length) == "") && (aSuggestions.length == 1)) {
                	oAutoSuggestControl.hideSuggestions();
                	return;
                }
            } 
        }
    }
    
    //provide suggestions to the control
    oAutoSuggestControl.autosuggest(aSuggestions);
};

/*===================================================================*/
if(!eXo.social) eXo.social = {};
if(!eXo.social.space) eXo.social.space = {};
eXo.social.space.UISpaceSearch = new UISpaceSearch();

File: portlet/space/src/main/java/social/portal/webui/component/space/UISpaceSearch.java
/*
 * Copyright (C) 2003-2009 eXo Platform SAS.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Affero General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see<http://www.gnu.org/licenses/>.
 */
package social.portal.webui.component.space;

import java.util.ArrayList;
import java.util.List;
import java.util.ResourceBundle;

import org.exoplatform.social.core.identity.ProfileFiler;
import org.exoplatform.social.space.Space;
import org.exoplatform.social.space.SpaceService;
import org.exoplatform.webui.application.WebuiRequestContext;
import org.exoplatform.webui.config.annotation.ComponentConfig;
import org.exoplatform.webui.config.annotation.EventConfig;
import org.exoplatform.webui.core.UIComponent;
import org.exoplatform.webui.core.lifecycle.UIFormLifecycle;
import org.exoplatform.webui.event.Event;
import org.exoplatform.webui.event.EventListener;
import org.exoplatform.webui.form.UIForm;
import org.exoplatform.webui.form.UIFormStringInput;

import social.portal.webui.component.UIManageInvitationSpaces;
import social.portal.webui.component.UIManageMySpaces;
import social.portal.webui.component.UIManagePendingSpaces;
import social.portal.webui.component.UIManagePublicSpaces;

/**
 * UIProfileUserSearch for search users in profile.
 * The search event should broadcast for the parent one to catch and
 * get searched Identity List from UIProfileUserSearch
 * Author : hanhvi
 *          hanhvq@gmail.com
 * Oct 28, 2009  
 */
@ComponentConfig(
  lifecycle = UIFormLifecycle.class,
  template = "app:/groovy/portal/webui/space/UISpaceSearch.gtmpl",
  events = {
    @EventConfig(listeners = UISpaceSearch.SearchActionListener.class) 
  }
)
public class UISpaceSearch extends UIForm {
  /** USER CONTACT. */
  final public static String SPACE_SEARCH = "SpaceSearch";
  /** SEARCH. */
  final public static String SEARCH = "Search";
  
  final static String ALL = "All";
  
  /** DEFAULT SPACE NAME SEARCH. */
  final public static String DEFAULT_SPACE_NAME_SEARCH = "Space name";
  final static String RIGHT_INPUT_PATTERN = "^[\\p{L}][\\p{L}._\\- \\d]+$";
  /** ADD PREFIX TO ENSURE ALWAY RIGHT THE PATTERN FOR CHECKING */
  final static String PREFIX_ADDED_FOR_CHECK = "PrefixAddedForCheck";
  
  SpaceService spaceService = null;
  private List<Space> spaceList = null;
  /** Selected character when search by alphabet */
  String selectedChar = null;
  String spaceNameSearch = null;
  
  public String getSpaceNameSearch() { return spaceNameSearch;}
  
  public void setSpaceNameSearch(String spaceNameSearch) { this.spaceNameSearch = spaceNameSearch;}

  /**
   * Constructor to initialize form fields
   * @throws Exception
   */
  public UISpaceSearch() throws Exception { 
    addUIFormInput(new UIFormStringInput(SPACE_SEARCH, null, DEFAULT_SPACE_NAME_SEARCH));
  }
  
  /**
   * Get all space for searching suggestion.
   * 
   * @return all space
   * @throws Exception 
   */
  public List<String> getAllSpaceName() throws Exception {
    List<String> allSpaceName = new ArrayList<String>();
//    SpaceService spaceService = getSpaceService();
//    List<Space> allSpace = spaceService.getAllSpaces();
    List<Space> allSpace = new ArrayList<Space>();
    UIComponent parent = getParent();
    if (parent instanceof UIManageMySpaces) allSpace = ((UIManageMySpaces)parent).getAllUserSpaces();
    if (parent instanceof UIManageInvitationSpaces) allSpace = ((UIManageInvitationSpaces)parent).getInvitationSpaces();
    if (parent instanceof UIManagePendingSpaces) allSpace = ((UIManagePendingSpaces)parent).getAllPendingSpaces();
    if (parent instanceof UIManagePublicSpaces) allSpace = ((UIManagePublicSpaces)parent).getAllPublicSpaces();
    
    for (Space space : allSpace) {
      allSpaceName.add(space.getName());
    }
    
    return allSpaceName;
  }
  
  /**
   * identityList setter
   * @param identityList
   */
  public void setSpaceList(List<Space> spaceList) {
    this.spaceList = spaceList;
  }
  
  /**
   * identityList getter
   * @return
   * @throws Exception 
   */
  public List<Space> getSpaceList() throws Exception {
    return spaceList;
  }

  public String getSelectedChar() { return selectedChar;}

  public void setSelectedChar(String selectedChar) { this.selectedChar = selectedChar;}
  
  /**
   * SearchActionListener
   * Get the space name from request and search spaces that have name like input space name.
   * Search space and set result into spaceList. 
   */
  static public class SearchActionListener extends EventListener<UISpaceSearch> {
    @Override
    public void execute(Event<UISpaceSearch> event) throws Exception {
      WebuiRequestContext ctx = event.getRequestContext();
      UISpaceSearch uiSpaceSearch = event.getSource();
      String charSearch = ctx.getRequestParameter(OBJECTID);
      SpaceService spaceService = uiSpaceSearch.getSpaceService();
      ResourceBundle resApp = ctx.getApplicationResourceBundle();
      String defaultSpaceName = resApp.getString(uiSpaceSearch.getId() + ".label.SpaceName");
      String spaceName = (((UIFormStringInput)uiSpaceSearch.getChild(UIFormStringInput.class)).getValue());
      if (spaceName != null) spaceName = spaceName.trim();
      
      spaceName = ((spaceName == null) || (spaceName.length() == 0) || spaceName.equals(defaultSpaceName)) ? "*" : spaceName;
      spaceName = (charSearch != null) ? charSearch : spaceName;
      spaceName = ((charSearch != null) && ALL.equals(charSearch)) ? "" : spaceName;
      
      if (charSearch != null) ((UIFormStringInput)uiSpaceSearch.getChildById(SPACE_SEARCH)).setValue(DEFAULT_SPACE_NAME_SEARCH);
      uiSpaceSearch.setSelectedChar(charSearch);
         
      if (charSearch == null) { // is not search by first character
        if (!isValidInput(spaceName)) {
          uiSpaceSearch.setSpaceList(new ArrayList<Space>());
        } else {
          spaceName = (spaceName.charAt(0) != '*') ? "*" + spaceName : spaceName;
          spaceName = (spaceName.charAt(spaceName.length()-1) != '*') ? spaceName += "*" : spaceName;
          spaceName = (spaceName.indexOf("*") >= 0) ? spaceName.replace("*", ".*") : spaceName;
          spaceName = (spaceName.indexOf("%") >= 0) ? spaceName.replace("%", ".*") : spaceName;
          List<Space> spaceSearchResult = spaceService.getSpacesByName(spaceName, false);
          uiSpaceSearch.setSpaceList(spaceSearchResult);
        }
      } else {
        List<Space> spaceSearchResult = spaceService.getSpacesByName(spaceName, true );
        uiSpaceSearch.setSpaceList(spaceSearchResult);  
      }
      
      uiSpaceSearch.setSpaceNameSearch(spaceName);
      
      Event<UIComponent> searchEvent = uiSpaceSearch.<UIComponent>getParent().createEvent(SEARCH, Event.Phase.DECODE, ctx);
      if (searchEvent != null) {
        searchEvent.broadcast();
      }
    }
    
    /**
     * Check input values is right or not follow regular expression.
     * @param input
     * @return
     */
    private boolean isValidInput(String input) {
      // Eliminate '*' and '%' character in string for checking
      String spacenameForCheck = input.replace("*", "");
      spacenameForCheck = spacenameForCheck.replace("%", "");
      // Make sure string for checking is started by alphabet character
      spacenameForCheck =  PREFIX_ADDED_FOR_CHECK + spacenameForCheck;
      if (!spacenameForCheck.matches(RIGHT_INPUT_PATTERN)) return false;
    
      return true;
    }
  }
  
  /**
   * get {@SpaceService}
   * @return spaceService
   */
  private SpaceService getSpaceService() {
    if (spaceService == null) {
      spaceService = getApplicationComponent(SpaceService.class);
    }
    return spaceService;
  }
  
}
